File: src/test/java/learnyouakotlin/end/java/JsonFormat.java
package learnyouakotlin.end.java;

import com.fasterxml.jackson.databind.JsonMappingException;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.node.ObjectNode;

import java.text.ParseException;
import java.util.List;
import java.util.Objects;
import java.util.stream.Collectors;

import static java.util.Spliterator.ORDERED;
import static java.util.Spliterators.spliterator;
import static java.util.stream.StreamSupport.stream;
import static learnyouakotlin.end.java.Json.*;

public class JsonFormat {
    public static JsonNode sessionAsJson(Session session) {
        return obj(
                prop("code", session.code.toString()),
                prop("title", session.title),
                prop("presenters", array(session.presenters, JsonFormat::presenterAsJson)));
    }

    public static Session sessionFromJson(JsonNode json) throws JsonMappingException {
        try {
            SessionCode code = SessionCode.parse(json.path("code").asText());
            String title = nonBlankText(json.path("title"));

            JsonNode authorsNode = json.path("presenters");
            List<Presenter> authors = stream(spliterator(authorsNode.elements(), authorsNode.size(), ORDERED), false)
                    .map(JsonFormat::presenterFromJson)
                    .collect(Collectors.toList());

            return new Session(code, title, authors);

        } catch (ParseException e) {
            throw new JsonMappingException(null, "failed to parse Session from JSON", e);
        }
    }

    private static ObjectNode presenterAsJson(Presenter p) {
        return obj(prop("name", p.name));
    }

    private static Presenter presenterFromJson(JsonNode authorNode) {
        return new Presenter(authorNode.path("name").asText());
    }

    private static String nonBlankText(JsonNode node) throws JsonMappingException {
        String text = node.asText();
        if (Objects.equals(text, "")) {
            throw new JsonMappingException(null, "missing or empty text");
        } else {
            return text;
        }
    }
}

File: src/test/java/learnyouakotlin/end/kotlin/JsonFormat.kt
package learnyouakotlin.end.kotlin

import com.fasterxml.jackson.databind.JsonMappingException
import com.fasterxml.jackson.databind.JsonNode
import learnyouakotlin.end.kotlin.Result.Failure
import learnyouakotlin.end.kotlin.Result.Success

fun Session.asJson() = obj(
    "code" of code.toString(),
    "title" of title,
    "presenters" of array(presenters, Presenter::asJson))

fun JsonNode.toSession() = apply(::Session,
    path("code").toSessionCode(),
    path("title").asNonblankText(),
    path("presenters").flatMap(JsonNode::toPresenter))


fun Presenter.asJson() = obj("name" of name)
fun JsonNode.toPresenter() = apply(::Presenter, path("name").asNonblankText())

fun JsonNode.toSessionCode() = this.asText()
    .let { SessionCode.parse(it) }
    ?.let { Success(it) }
    ?: jsonFailure("could not parse ${asText()} as SessionCode")


fun JsonNode.asNonblankText() = asText().let {
    when {
        it.isBlank() -> jsonFailure("blank text is invalid")
        else -> Result.Success(it)
    }
}

private fun <T> JsonNode.map(transform: (JsonNode) -> T) = elements().asSequence().map(transform).toList()
private fun <T> JsonNode.flatMap(transform: (JsonNode) -> Result<T>) = map(transform).flatten()

private fun jsonFailure(message: String) = Failure(JsonMappingException(null, message))

